### PR TITLE
Isolation: Refactor resetting RCReadCheckTS internal flag 

### LINKED ISSUE #37226: txn: allow larger initial chunk size for read check-ts for read-consistency
## Enhancement

When the read check-ts [feature](https://docs.pingcap.com/zh/tidb/dev/system-variables#tidb_rc_read_check_ts%E4%BB%8E-v600-%E7%89%88%E6%9C%AC%E5%BC%80%E5%A7%8B%E5%BC%95%E5%85%A5) is used, there's a constraint that the statement retry could work only when conflicts are detected and there're no result packets are responded to the client yet.

However, the init chunk size has a limit [upper bound](https://docs.pingcap.com/tidb/dev/system-variables#tidb_init_chunk_size) which is 32, which means the select statement may respond error to the client if the total result row number is larger than 32 and there's possible conflict after the first chunk is responded. As a result, the client may receive errors and have to process these errors if the read check-ts feature is enabled and the whole result row number is large, even the conflict happens rarely.

An enhancement for the read check-ts feature is to allow the init chunk size to be adjusted to a larger value like 1024 so the client does not need to handle errors when the conflict happens rarely and the select result row number is large.

### PR DESCRIPTION
**Pull request by TonsnakeLin**
Signed-off-by: TonsnakeLin <lpbgytong@163.com>

### What problem does this PR solve?

Issue Number: Ref #37226

Problem Summary:

### What is changed and how it works?
In the past, TiDB reset the internal flag `PessimisticRCTxnContextProvider::latestOracleTSValid` in function `(p *PessimisticRCTxnContextProvider) handleAfterQueryError` and `(p *PessimisticRCTxnContextProvider) handleAfterPessimisticLockError`. This PR moved it to `(p *PessimisticRCTxnContextProvider) OnStmtRetry`, so that we can statistic the retry count because of `RCCheckTS` easily. 

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by TonsnakeLin** [comment]
@cfzjywxk @ekexium I unified resetting internal flags of `RCReadCheckTS` and `RCWriteCheckTS`, so that we can statistic it at same place, Thanks for reviewing this PR.

File: server/tidb_test.go [hunk 1]
```diff
@@ -2632,6 +2633,64 @@ func TestRcReadCheckTSConflict(t *testing.T) {
...
+	tk2.MustExec("update t1 set id3 = id3 + 1 where id1 = 1")
+	require.NoError(t, cc.handleStmtPrepare(ctx, "select * from t1 where id1 = 1"))
+	require.NoError(t, cc.handleStmtExecute(ctx, []byte{0x1, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0}))
+	tk.MustExec("commit")
+	count, ok = se.Value(sessiontxn.CallOnStmtRetryCount).(int)
+	require.Equal(t, true, ok)
+	require.Equal(t, 1, count)
+
+	tk.MustExec("drop table t1")
+	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/sessiontxn/isolation/CallOnStmtRetry"))
```
Comment (cfzjywxk) [inline server/tidb_test.go:2691]: Better to disable the failpoint in a `defer func` to avoid affecting other concurrent cases if an error happens.

**Comment by ekexium** [comment]
The change itself looks good. I'm jsut wondering how you will tell the reason of a statement retry.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Comment by TonsnakeLin** [comment]
> The change itself looks good. I'm jsut wondering how you will tell the reason of a statement retry.

I mentioned it at PR #37226

File: sessiontxn/failpoint.go [hunk 2]
```diff
@@ -52,6 +52,9 @@ var TsoWaitCount stringutil.StringerStr = "tsoWaitCount"
 // TsoUseConstantCount is the key for constant tso counter
 var TsoUseConstantCount stringutil.StringerStr = "tsoUseConstantCount"
 
+// CallOnStmtRetryCount is the key for recording calling OnStmtRetry at RC isolation level
+var CallOnStmtRetryCount stringutil.StringerStr = "callOnStmtRetryCount"
+
 // AssertLockErr is used to record the lock errors we encountered
 // Only for test
 var AssertLockErr stringutil.StringerStr = "assertLockError"
```
Comment (hawkingrei) [inline sessiontxn/failpoint.go:60]: Caen they be defined as const variables?

**Comment by ekexium** [comment]
> 

I See. But the write conflict error is unavailable in `OnStmtRetry`, how could you tell the reason?

**Comment by TonsnakeLin** [comment]
> > 
> 
> I See. But the write conflict error is unavailable in `OnStmtRetry`, how could you tell the reason?

Yes, I I realized this. Maybe it's not suitable to to it in function `OnStmtRetry`, or we can add a new flag in `RCContextProvider`, but it's ugly.
IMO, it maybe still collect the `RCCheckTS WriteConflict` in function `(p *PessimisticRCTxnContextProvider) handleAfterQueryError and (p *PessimisticRCTxnContextProvider) handleAfterPessimisticLockError` as I said in #37226.
But it doesn't affect this refactoring.

**Review by ekexium — APPROVED** [review]

**Comment by TonsnakeLin** [comment]
@cfzjywxk please merge it ,thank you.

**Pull request merged**